File: python/04_error_handling.py
"""
Cynco API Quickstart -- 04: Error Handling

Demonstrates how to handle every error type the Cynco API can return:
- 400 Bad Request
- 401 Unauthorized
- 403 Forbidden
- 404 Not Found
- 409 Conflict
- 422 Validation Error
- 429 Rate Limited
- 5xx Server Error

Also includes a production-ready retry utility with exponential backoff.

Run: python 04_error_handling.py
"""

import os
import sys
import time
from typing import Any, Callable, TypeVar

import requests

API_KEY: str = os.environ.get("CYNCO_API_KEY", "")
BASE_URL: str = "https://app.cynco.io/api/v1"

if not API_KEY:
    print("Missing CYNCO_API_KEY environment variable.")
    print("Set it with: export CYNCO_API_KEY='cak_your_api_key_here'")
    sys.exit(1)

headers: dict[str, str] = {
    "Authorization": f"Bearer {API_KEY}",
}

T = TypeVar("T")


class CyncoAPIError(Exception):
    """Base error for Cynco API responses."""

    def __init__(
        self,
        status_code: int,
        code: str,
        message: str,
        request_id: str | None = None,
        details: list[dict[str, str]] | None = None,
    ) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.code = code
        self.request_id = request_id
        self.details = details or []


def handle_response(response: requests.Response) -> dict:
    """
    Parse a Cynco API response, raising typed exceptions for errors.
    Use this as a central response handler in your application.
    """
    # 2xx responses are successful.
    if response.ok:
        return response.json()

    # Parse the error body.
    try:
        body: dict = response.json()
        error: dict = body.get("error", {})
        meta: dict = body.get("meta", {})
    except ValueError:
        # Non-JSON error response (rare).
        raise CyncoAPIError(
            status_code=response.status_code,
            code="UNKNOWN",
            message=f"HTTP {response.status_code}: {response.text}",
        )

    raise CyncoAPIError(
        status_code=response.status_code,
        code=error.get("code", "UNKNOWN"),
        message=error.get("message", "Unknown error"),
        request_id=meta.get("requestId"),
        details=error.get("details"),
    )


def demonstrate_error_handling() -> None:
    """Show how to catch and handle different error types."""
    print("=== Error Handling Examples ===\n")

    # 1. Not Found (404) -- fetching a nonexistent resource.
    print("--- 404 Not Found ---")
    try:
        response = requests.get(
            f"{BASE_URL}/customers/cus_nonexistent",
            headers=headers,
            timeout=30,
        )
        result = handle_response(response)
        print(f"Customer: {result['data']['name']}")
    except CyncoAPIError as e:
        if e.status_code == 404:
            print(f"  Resource not found: {e.message}")
            print(f"  Request ID: {e.request_id}")
        else:
            print(f"  Unexpected error ({e.status_code}): {e.message}")
    print()

    # 2. Validation Error (422) -- sending invalid data.
    print("--- 422 Validation Error ---")
    try:
        response = requests.post(
            f"{BASE_URL}/customers",
            headers={**headers, "Content-Type": "application/json"},
            json={
                "name": "",  # Name is required
                "email": "not-an-email",  # Invalid email format
            },
            timeout=30,
        )
        result = handle_response(response)
        print(f"Customer: {result['data']['name']}")
    except CyncoAPIError as e:
        if e.status_code == 422:
            print(f"  Validation failed: {e.message}")
            for detail in e.details:
                print(f"    {detail['field']}: {detail['message']}")
        else:
            print(f"  Unexpected error ({e.status_code}): {e.message}")
    print()

    # 3. Rate Limit (429) -- too many requests.
    print("--- 429 Rate Limit ---")
    print("  (Simulating rate limit handling)")
    print("  On a real 429, the response includes:")
    print("    Retry-After: <seconds until you can retry>")
    print("    X-RateLimit-Limit: <your tier limit>")
    print("    X-RateLimit-Remaining: 0")
    print("  Use the retry utility below to handle this automatically.")
    print()


def with_retry(
    fn: Callable[[], T],
    max_retries: int = 3,
    base_delay: float = 1.0,
) -> T:
    """
    Retry a function with exponential backoff.
    Retries on rate limits (429) and server errors (5xx).
    Does not retry client errors (4xx except 429).
    """
    for attempt in range(max_retries + 1):
        try:
            return fn()
        except CyncoAPIError as e:
            is_last_attempt = attempt == max_retries

            if e.status_code == 429:
                # Rate limited -- use the Retry-After hint if available.
                if is_last_attempt:
                    raise
                delay = base_delay * (2 ** attempt)
                print(f"  Rate limited. Retrying in {delay:.1f}s...")
                time.sleep(delay)

            elif e.status_code >= 500:
                # Server error -- retry with backoff.
                if is_last_attempt:
                    raise
                delay = base_delay * (2 ** attempt)
                print(f"  Server error ({e.status_code}). Retrying in {delay:.1f}s...")
                time.sleep(delay)

            else:
                # Client error (4xx except 429) -- don't retry.
                raise

    # Unreachable, but satisfies the type checker.
    raise RuntimeError("Unreachable")


def demonstrate_retry() -> None:
    """Show the retry utility in action."""
    print("=== Retry with Backoff ===\n")

    def fetch_invoices() -> dict:
        response = requests.get(
            f"{BASE_URL}/invoices",
            headers=headers,
            params={"limit": 5},
            timeout=30,
        )
        return handle_response(response)

    try:
        result = with_retry(fetch_invoices, max_retries=3, base_delay=1.0)
        invoice_count = len(result["data"])
        print(f"Fetched {invoice_count} invoices with retry support.\n")
    except CyncoAPIError as e:
        print(f"All retries exhausted: {e.message}")
    except requests.RequestException as e:
        print(f"Network error: {e}")


def main() -> None:
    demonstrate_error_handling()
    demonstrate_retry()


if __name__ == "__main__":
    main()

File: typescript/src/04-error-handling.ts
/**
 * Cynco API Quickstart -- 04: Error Handling
 *
 * Demonstrates how to handle every error type the SDK can throw:
 * - ValidationError (422): invalid input data
 * - AuthenticationError (401): bad or missing API key
 * - PermissionError (403): key lacks required permissions
 * - NotFoundError (404): resource doesn't exist
 * - ConflictError (409): state conflict or duplicate
 * - RateLimitError (429): too many requests
 * - APIError (5xx): server-side errors
 *
 * Run: npm run example:errors
 */

import Cynco from "@cynco/sdk";

const cynco = new Cynco({
  apiKey: process.env.CYNCO_API_KEY,
});

/**
 * Comprehensive error handling with the SDK's typed errors.
 * Each error type has specific properties you can inspect.
 */
async function handleAllErrorTypes(): Promise<void> {
  console.log("=== Error Handling Examples ===\n");

  try {
    // This will fail if the customer ID doesn't exist.
    const customer = await cynco.customers.retrieve("cus_nonexistent");
    console.log("Customer:", customer.name);
  } catch (error: unknown) {
    if (error instanceof Cynco.NotFoundError) {
      // 404 -- the resource doesn't exist.
      console.log("[NotFoundError]");
      console.log(`  Message: ${error.message}`);
      console.log(`  Status:  ${error.status}`);
      console.log(`  Code:    ${error.code}`);
      console.log(`  Request: ${error.requestId}\n`);
    } else if (error instanceof Cynco.ValidationError) {
      // 422 -- input data failed validation.
      // The `details` array tells you exactly which fields are wrong.
      console.log("[ValidationError]");
      console.log(`  Message: ${error.message}`);
      for (const detail of error.details) {
        console.log(`  Field "${detail.field}": ${detail.message}`);
      }
      console.log();
    } else if (error instanceof Cynco.AuthenticationError) {
      // 401 -- the API key is invalid or missing.
      console.log("[AuthenticationError]");
      console.log(`  Message: ${error.message}`);
      console.log("  Action: Check your CYNCO_API_KEY environment variable.\n");
    } else if (error instanceof Cynco.PermissionError) {
      // 403 -- the API key doesn't have permission for this operation.
      console.log("[PermissionError]");
      console.log(`  Message: ${error.message}`);
      console.log(
        "  Action: Check key permissions in Settings > API Keys.\n"
      );
    } else if (error instanceof Cynco.RateLimitError) {
      // 429 -- too many requests.
      // The `retryAfter` property tells you when to retry (in seconds).
      console.log("[RateLimitError]");
      console.log(`  Message:     ${error.message}`);
      console.log(`  Retry after: ${error.retryAfter}s`);
      console.log(`  Limit:       ${error.limit} req/min\n`);
    } else if (error instanceof Cynco.ConflictError) {
      // 409 -- state conflict (e.g., trying to void an already-voided invoice).
      console.log("[ConflictError]");
      console.log(`  Message: ${error.message}\n`);
    } else if (error instanceof Cynco.APIError) {
      // 5xx -- server error. Safe to retry with backoff.
      console.log("[APIError]");
      console.log(`  Status:  ${error.status}`);
      console.log(`  Message: ${error.message}`);
      console.log(`  Request: ${error.requestId}\n`);
    } else {
      // Network errors, timeouts, etc.
      console.error("[Unknown error]", error);
    }
  }
}

/**
 * Retry logic with exponential backoff.
 * Useful for handling transient errors (429, 5xx) in production.
 */
async function withRetry<T>(
  fn: () => Promise<T>,
  options: { maxRetries?: number; baseDelay?: number } = {}
): Promise<T> {
  const { maxRetries = 3, baseDelay = 1000 } = options;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error: unknown) {
      const isLastAttempt = attempt === maxRetries;

      // Only retry on rate limits and server errors.
      if (error instanceof Cynco.RateLimitError) {
        if (isLastAttempt) throw error;
        const delay = error.retryAfter * 1000;
        console.log(`Rate limited. Retrying in ${error.retryAfter}s...`);
        await sleep(delay);
      } else if (
        error instanceof Cynco.APIError &&
        error.status >= 500
      ) {
        if (isLastAttempt) throw error;
        // Exponential backoff: 1s, 2s, 4s
        const delay = baseDelay * Math.pow(2, attempt);
        console.log(
          `Server error (${error.status}). Retrying in ${delay}ms...`
        );
        await sleep(delay);
      } else {
        // Don't retry client errors (4xx except 429).
        throw error;
      }
    }
  }

  // TypeScript requires this -- the loop always returns or throws above.
  throw new Error("Unreachable");
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main(): Promise<void> {
  await handleAllErrorTypes();

  // Demonstrate retry logic.
  console.log("=== Retry with Backoff ===\n");

  try {
    const invoices = await withRetry(
      () => cynco.invoices.list({ limit: 5 }),
      { maxRetries: 3, baseDelay: 1000 }
    );
    console.log(
      `Fetched ${invoices.data.length} invoices with retry support.\n`
    );
  } catch (error: unknown) {
    console.error("All retries exhausted:", error);
  }
}

main().catch((error: unknown) => {
  console.error("Error handling example failed:", error);
  process.exit(1);
});
